src/main/scala/is/hail/utils/richUtils/RichRDDByteArray.scala

```diff
@@ -8,30 +8,48 @@ import org.apache.spark.rdd.RDD
 import scala.reflect.ClassTag
 
 class RichRDDByteArray(val r: RDD[Array[Byte]]) extends AnyVal {
-  def saveFromByteArrays(filename: String, tmpDir: String, header: Option[Array[Byte]] = None, deleteTmpFiles: Boolean = true) {
+  def saveFromByteArrays(filename: String, tmpDir: String, header: Option[Array[Byte]] = None,
+    deleteTmpFiles: Boolean = true, parallelWrite: Boolean = false) {
     val nullWritableClassTag = implicitly[ClassTag[NullWritable]]
     val bytesClassTag = implicitly[ClassTag[BytesOnlyWritable]]
     val hConf = r.sparkContext.hadoopConfiguration
 
-    val tmpFileName = hConf.getTemporaryFile(tmpDir)
+    hConf.delete(filename, recursive = true) // overwriting by default
 
-    val rMapped = r.mapPartitions { iter =>
+    val parallelOutputPath = if (parallelWrite) filename else hConf.getTemporaryFile(tmpDir)
+
+    val rWithHeader = header.map { h =>
+      if (r.partitions.length == 0)
+        r.sparkContext.parallelize(List(h), numSlices = 1)
+      else if (parallelWrite)
+        r.mapPartitions { it => Iterator(h) ++ it }
+      else
+        r.mapPartitionsWithIndex { case (i, it) =>
+          if (i == 0)
+            Iterator(h) ++ it
+          else
+            it
+        }
+    }.getOrElse(r)
+
+    val rMappedWithHeader = rWithHeader.mapPartitions { iter =>
       val bw = new BytesOnlyWritable()
       iter.map { bb =>
         bw.set(new BytesWritable(bb))
         (NullWritable.get(), bw)
       }
     }
 
-    RDD.rddToPairRDDFunctions(rMapped)(nullWritableClassTag, bytesClassTag, null)
-      .saveAsHadoopFile[ByteArrayOutputFormat](tmpFileName)
+    RDD.rddToPairRDDFunctions(rMappedWithHeader)(nullWritableClassTag, bytesClassTag, null)
+      .saveAsHadoopFile[ByteArrayOutputFormat](parallelOutputPath)
```

Comment:
The explicit `rddToPairRDDFunctions` is unnecessary, because it is implicit. Let's clean up this code while we're in here:
```scala
rMappedWithHeader.saveAsHadoopFile(parallelOutputPath)
```

Reply:
I'm not sure if the type parameter is required or not, it seems that it might be required.